modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java

```diff
@@ -1208,12 +1208,33 @@ public IgniteSemaphore semaphore(final String name, @Nullable AtomicConfiguratio
 
                 if (val == null && !create)
                     return null;
-
+           
                 AtomicDataStructureValue retVal = (val == null ? new GridCacheSemaphoreState(cnt,
                     new HashMap<UUID, Integer>(), failoverSafe, ctx.discovery().gridStartTime()) : null);
 
                 GridCacheSemaphoreEx sem0 = new GridCacheSemaphoreImpl(name, key, cache);
 
+                //check Cluster state against semaphore state
+                if(val!=null){ 
+                    ctx.closure().callLocalSafe(
+                        new Callable<Object>() {
+                            @Override
+                            public Object call() throws Exception {
+                                GridCacheSemaphoreState semState=(GridCacheSemaphoreState)val;
+
+                                for (UUID nodeId : semState.getWaiters().keySet()) {
+                                    ClusterNode node = ctx.cluster().get().node(nodeId);
+                                    if (node == null) {
+                                        sem0.onNodeRemoved(nodeId);
```

Comment:
Same goes here, this method onNodeRemoved() executes a new transaction, we shouldn't start a new transaction from within a transaction.